capsules/src/usb/cdc.rs

```diff
@@ -229,24 +252,61 @@ impl<'a, U: hil::usb::UsbController<'a>> hil::usb::Client<'a> for CdcAcm<'a, U>
             .endpoint_set_out_buffer(ENDPOINT_OUT_NUM, self.buffer(ENDPOINT_OUT_NUM));
         self.controller()
             .endpoint_out_enable(TransferType::Bulk, ENDPOINT_OUT_NUM);
+
+        self.state.set(State::Enabled);
     }
 
     fn attach(&'a self) {
         self.client_ctrl.attach();
+        self.state.set(State::Attached);
     }
 
     fn bus_reset(&'a self) {
-        // This signals enumeration has finished, and at this point we can ask
-        // for IN transfers if needed.
-        self.initialized.set(true);
-
-        if self.tx_buffer.is_some() {
-            self.controller().endpoint_resume_in(ENDPOINT_IN_NUM);
-        }
+        // We take a bus reset to mean the enumeration has finished.
+        self.state.set(State::Enumerated);
     }
 
-    /// Handle a Control Setup transaction
+    /// Handle a Control Setup transaction.
+    ///
+    /// CDC uses special values here, and we can use these to know when a CDC
+    /// client is connected or not.
     fn ctrl_setup(&'a self, endpoint: usize) -> hil::usb::CtrlSetupResult {
+        descriptors::SetupData::get(&self.client_ctrl.ctrl_buffer.buf).map(|setup_data| {
+            let b_request = setup_data.request_code;
+            let value = setup_data.value;
+
+            // Match on the two CDC control messages we care about:
+            // - `0x22`: SET_LINE_CONTROL_STATE
+            // - `0x23`: SEND_BREAK
+            match b_request {
+                0x22 => {
+                    // This message seems to come with different values. We
+                    // don't care about the actual value's meaning, except for
+                    // value=0 which seems to bookend when the CDC client is
+                    // actually attached.
+                    if value == 0 {
+                        // On Linux, this seems to be a good signal of both when
+                        // a client connects and disconnects. So, based on our
+                        // current state, we can update our state.
+                        if self.state.get() != State::Connected {
+                            // We weren't previously connected so this must mean
+                            // a client connected.
+                            self.state.set(State::Connecting);
+                        } else {
+                            // We were connected, so disconnect.
+                            self.state.set(State::Enumerated)
+                        }
+                    }
+                }
+                0x23 => {
+                    // On Mac, we seem to get the SEND_BREAK to signal that a
+                    // client disconnects.
+                    self.state.set(State::Enumerated)
+                }
+                _ => {}
```

Comment:
Is there a bunch of unrelated traffic, or maybe future debugging efforts will be grateful for a debug print `Ignoring unknown CDC control message: {}`

Reply:
I don't think there are messages generally worth caring about, and it might vary based on implementations/operating systems. But in any case I don't think anyone would care about what the messages are.